tests/pytests/unit/states/test_jboss7.py

```diff
@@ -0,0 +1,589 @@
+import pytest
+import salt.states.jboss7 as jboss7
+from salt.exceptions import CommandExecutionError
+from tests.support.mock import MagicMock, patch
+
+
+@pytest.fixture
+def configure_loader_modules():
+    return {
+        jboss7: {
+            "__salt__": {
+                "jboss7.read_datasource": MagicMock(),
+                "jboss7.create_datasource": MagicMock(),
+                "jboss7.update_datasource": MagicMock(),
+                "jboss7.remove_datasource": MagicMock(),
+                "jboss7.read_simple_binding": MagicMock(),
+                "jboss7.create_simple_binding": MagicMock(),
+                "jboss7.update_simple_binding": MagicMock(),
+                "jboss7.undeploy": MagicMock(),
+                "jboss7.deploy": MagicMock,
+                "file.get_managed": MagicMock,
+                "file.manage_file": MagicMock,
+                "jboss7.list_deployments": MagicMock,
+            },
+            "__env__": "base",
+        }
+    }
+
+
+@pytest.fixture
+def list_deployments():
+    def list_deps(*args):
+        return ["some_artifact"]
+
+    return list_deps
+
+
+@pytest.fixture
+def list_deployments_other():
+    def list_deps_other(*args):
+        return ["some_other_artifact"]
+
+    return list_deps_other
+
+
+@pytest.fixture
+def file_get_managed():
+    return "sfn", "hash", ""
+
+
+@pytest.fixture
+def file_manage_file_true():
+    def file_manage_file(**kwargs):
+        return {"result": True, "changes": True}
+
+    return file_manage_file
+
+
+@pytest.fixture
+def file_manage_file_false():
+    def file_manage_file(**kwargs):
+        return {"result": True, "changes": False}
+
+    return file_manage_file
+
+
+def test_should_not_redeploy_unchanged(
+    list_deployments, file_get_managed, file_manage_file_false
+):
+    parameters = {
+        "target_file": "some_artifact",
+        "undeploy_force": False,
+        "undeploy": "some_artifact",
+        "source": "some_artifact_on_master",
+    }
+    jboss_conf = {"cli_path": "somewhere", "controller": "some_controller"}
+    jboss7_undeploy_mock = MagicMock()
+    jboss7_deploy_mock = MagicMock()
+    file_get_managed = MagicMock(side_effect=file_get_managed)
```

Comment:
**suggestion(non-blocking)**: can shift this type of mock into the fixtures, e.g.

```
@pytest.fixture
def file_get_managed():
    jboss7.__salt__["file.get_managed"] = create_autospec("salt.modules.file.get_managed", side_effect=["sfn", "hash", ""])
```

if a patch is necessary, can do 

```
with patch(...):
    yield  # optional data to yield if it's necessary in the test
```